Download Sample Datasets

fm3 opened the pull request:
### URL of deployed dev instance (used for testing):
- https://demodataset.webknossos.xyz

### Steps to test:
- new routes `GET http://localhost:9000/data/datasets/sample/Connectomics_Department?token=secretScmBoyToken` and `POST http://localhost:9000/data/datasets/sample/Connectomics_Department/Sample_Cremi_dsA_plus/download?token=secretScmBoyToken`
- corresponding view can be found next to upload dataset (“or add sample dataset”)


### Issues:
- fixes #3642 

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- ~~[ ] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable~~
- [x] Updated [documentation](../blob/master/docs) if applicable
- [x] Needs datastore update after deployment
- [x] Ready for review

daniel-wer commented:
@fm3 I added a first version of a sample datasets modal that shows the available sample datasets and allows to add them to webKnossos. To make it easy to test, I added a button in the dataset view.

I noticed that it seems like the downloading status is never reported. When triggering the download, the status remains "available" for some seconds and then switches to "present" directly. Or maybe the downloading state is only there very briefly (I'm polling every 500ms), ideally it would be set directly after triggering the download :)

Threre's still stuff missing in the frontend:
- [ ] Decide where exactly the option to add sample datasets should be available
- [ ] Choose the correct datastore (or datastores?) to download from (hardcoded as localhost:9000 for now).

fm3 commented:
Thanks @daniel-wer !
The modal looks good :)
I found the issue causing the “downloading” state to not be reported. That should be fixed now.
Another thing we should consider: if the download fails in the back-end, the status goes back to available (not sure if setting it to failed would really work well, raising the question when that should be resetted). Could the frontend interpret repeated “available”s after the download request was started as “failed”, until the page is reloaded? I’m not sure what behavior I’d prefer here, maybe we can talk about that.

Also, if the first request fails, the download wasn’t started (there are a couple of checks first), which is also not reflected in the modal’s behavior right now, it will be polling indefinitely.

daniel-wer commented:
> Another thing we should consider: if the download fails in the back-end, the status goes back to available (not sure if setting it to failed would really work well, raising the question when that should be resetted). Could the frontend interpret repeated “available”s after the download request was started as “failed”, until the page is reloaded? I’m not sure what behavior I’d prefer here, maybe we can talk about that.

Is it guaranteed, that when requesting the sample dataset status after triggering the download of dataset X (and having waited for the response), that the status for X === "available" if and only if the download request failed or could it be that the status for X is STILL available, because the internal data structures were not updated yet?
If the former is guaranteed, then this should be pretty easy, otherwise it'll probably get a bit dirty.

> Also, if the first request fails, the download wasn’t started (there are a couple of checks first), which is also not reflected in the modal’s behavior right now, it will be polling indefinitely.

I fixed that :)

fm3 commented:
> Is it guaranteed, that when requesting the sample dataset status after triggering the download of dataset X (and having waited for the response), that the status for X === "available" if and only if the download request failed

I shoud think so, yes.

> I fixed that :)

Cool, thanks!

fm3 commented:
note to self, todo: add deterministic sorting to dataStoreDAO.findAll

daniel-wer commented:
I added the option to add Sample Datasets to a couple of spots (as we discussed):

- A link at the bottom of the Add Dataset view, which allows to add sample datasets even when there are existing datasets already.
- As a placeholder for the dataset list if there are no datasets, see below:

![no-datasets](https://user-images.githubusercontent.com/1702075/53099821-e47df480-3526-11e9-9e26-4dfe007c7bc5.png)

- As part of the onboarding flow - Add Dataset step. I've reworked this step quite a bit to incorporate the sample dataset option (which is the most prominent option by design as it is the easiest), see below:

![onboarding-datasets](https://user-images.githubusercontent.com/1702075/53099868-037c8680-3527-11e9-824e-89ce40484218.png)


I think this is ready for a first review round. Any feedback regarding the design, copy and functionality appreciated :)

daniel-wer commented:
Note that the current backend version expects `http://localhost/cremi.zip` to be a zipped dataset for testing.
We'll have to think about where we want to host the sample datasets (and which).

fm3 commented:
the docs link to `https://static.webknossos.org/data/e2006_wkw.zip`. I think, the cremi one actually also makes sense (small, includes segmentation)

philippotto commented on frontend/javascripts/dashboard/dataset/sample_datasets_modal.js:
```diff
@@ -0,0 +1,146 @@
+
+  useInterval(fetchDatasets, pendingDatasets.length ? 1000 : null);
+
+  const handleSampleDatasetDownload = async (name: string) => {
```
I'm wondering whether it makes sense to move `handleSampleDatasetDownload`, the `pendingDatasets` definition and the `useDatastores` call into the `useSampleDatasets` hook. At the moment, it feels a bit weird that `failedDatasets` are provided by the `useSampleDatasets` hook, but the `pendingDatasets` have to be maintained at an higher level.
`handleSampleDatasetDownload` would be another return argument of `useSampleDatasets`.

This would clean up the view logic here a bit, but obviously the `useSampleDatasets` hook would get longer. I'd argue, that the abstraction level is more "even", though. What do you think?

daniel-wer replied:
I struggled with that as well, thinking along the same lines ^^ I'll give this option a try and see what I can do to clean this up a little bit :)

philippotto commented on frontend/javascripts/dashboard/dataset/sample_datasets_modal.js:
```diff
@@ -0,0 +1,146 @@
+  organizationName: string,
+};
+
+function useDatastores(): [Array<APIDataStore>] {
```
I imagine that these ten lines of code will be useful quite a lot in the future (independent of entity type, of course). We don't have to do it now, but I'd think that something like `const datastores = useAsyncGet(getDataStores, [])` would be a good abstraction (naming is debatable).

daniel-wer replied:
I created a generic `useFetch` method and included a third parameter for the dependencies (for useEffect) :)

fm3 commented on frontend/javascripts/dashboard/dataset/sample_datasets_modal.js:
```diff
@@ -0,0 +1,146 @@
+    const sampleDatasets = await getSampleDatasets(datastore.url, organizationName);
+    setDatasets(sampleDatasets);
+
+    // Datasets that were pending, but are now available again, failed to download
```
Yes, since the download is running in the backend, it is not easy to communicate a “failed” state (it would have to expire). Hence it simply turns back to available. And the front-end treats that as failed for this session.

daniel-wer commented on frontend/javascripts/dashboard/dataset/sample_datasets_modal.js:
```diff
@@ -0,0 +1,146 @@
+
+const SampleDatasetsModal = ({ destroy, onOk, organizationName }: Props) => {
+  const [pendingDatasets, setPendingDatasets] = useState([]);
+  const [datastores] = useDatastores();
```
Yes it absolutely could (and probably should). At first I wanted to maintain the hooks pattern of returning [value, valueSetter], but that didn't really work out ^^

normanrz commented:
- https://static.webknossos.org/data/e2006_wkw.zip
Raw SBEM data and segmentation (sample cutout)
Connectomic reconstruction of the inner plexiform layer in the mouse retina
M Helmstaedter, KL Briggman, S Turaga, V Jain, HS Seung, W Denk.
Nature. 08 August 2013. https://doi.org/10.1038/nature12346

- https://static.webknossos.org/data/FD0144_wkw.zip
Raw SBEM data and segmentation (sample cutout)
FluoEM, virtual labeling of axons in three-dimensional electron microscopy data for long-range connectomics
F Drawitsch, A Karimi, KM Boergens, M Helmstaedter.
eLife. 14 August 2018. https://doi.org/10.7554/eLife.38976

- https://static.webknossos.org/data/MPRAGE_250um.zip
MRI data
T1-weighted in vivo human whole brain MRI dataset with an ultrahigh isotropic resolution of 250 μm
F Lüsebrink, A Sciarra, H Mattern, R Yakupov, O Speck
Scientific Data. 14 March 2017. https://doi.org/10.1038/sdata.2017.32

fm3 commented:
@daniel-wer I added the three datasets from above. Could you display the description in the front-end? I added it as `description` in the json. It contains line-breaks, not sure how that is best handled.

daniel-wer commented:
@fm3 I took care of the descriptions :)
During testing of the sample datasets, I noticed that the `Sample_FD0144_wkw` dataset cannot be openend, because no resolutions are sent as part of the dataset request. The datasource config (in the Edit view) does contain resolutions, though, so I'm not sure what the problem is. Maybe you could have a look at that?

fm3 commented:
There’s two layers named “color_2” in the datasource-properties.json in FD0144_wkw.zip. @normanrz could you update it? I suppose one of them should be “color_3”. Error handling for this is not great, but I don’t know a quick way to fix that. Created #3845 to track.

```
backend: 2019-02-28 13:33:59,110 [ERROR] models.binary.DataSetDataLayerDAO - SQL Error: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "dataset_layers_pkey"
backend:   Detail: Key (_dataset, name)=(5c77d537640100b601f3b0b5, color_2) already exists.
```

fm3 commented:
Thanks for fixing that @normanrz! one last thing: the real bounding box for that dataset (FD0144_wkw) seems to be smaller (I don’t see data for z>=164) so to optimize the user experience we might want to add that to the datasource-properties.json. Shouldn’t block the PR, though

normanrz commented:
fixed

fm3 commented:
Cool, thanks! I updated the docs. If you have no objections @philippotto @normanrz I suggest we can merge this today

philippotto commented on docs/getting_started.md:
```diff
@@ -70,12 +69,12 @@ However, webKnossos includes fine-grained permissions to assign datasets to grou
-Identify a dataset that your interested in and click on `Start Skeleton Tracing` to create a new skeleton annotation. 
-webKnossos will launch the main annotation screen allowing you to navigate your dataset and place markers to reconstruct skeletons. 
+To get started with your first annotation, navigate to the `Datasets` tab on your [dashboard](./dashboard.md).
+Identify a dataset that your interested in and click on `Start Skeleton Tracing` to create a new skeleton annotation.
```
I know that you only changed the whitespace here, but it should be `you are` :see_no_evil:

philippotto commented:
Docs look good to me :+1:
